include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+       *     operations tracked by the tape. If the @p keep flag is set to
+       *     <tt>true</tt> then these represent precisely the point about which
+       *     the function derivatives are to be computed. If the @p keep flag is
+       *     set to <tt>false</tt> then these only represent dummy values, and
+       *     the point at which the function derivatives are to be computed must
+       *     be set by calling set_independent_variables() again.
+       *   - Extraction of a set of independent variables of auto-differentiable
+       *     type using get_sensitive_variables(). These are then tracked during
+       *     later computations.
+       *   - Defining the dependent variables via register_dependent_variable()
+       *     or register_dependent_variables(). These are the functions that
+       *     will be differentiated with respect to the independent variables.
+       *
+       * @param[in] tape_index The index of the tape to be written
+       * @param[in] overwrite_tape Express whether tapes are allowed to be
+       * overwritten. If <tt>true</tt> then any existing tape with a given
+       * @p tape_index will be destroyed and a new tape traced over it.
+       * @param[in] keep_values Determines whether the numerical values of all
+       * independent variables are recorded in the tape buffer. If true, then
+       * tape can be immediately used to perform computations after recording
+       * is complete.
+       *
+       * @note During the recording phase, no value(), gradient(), hessian(),
+       * or jacobian() operations can be performed.
+       *
+       * @note The chosen tape index must be greater than
+       * types::invalid_tape_index and less than types::max_tape_index.
+       */
+      bool
+      enable_record_sensitivities(const types::tape_index &tape_index,
+                                  const bool overwrite_tape = false,
+                                  const bool keep_values    = true);
+
+      /**
+       * Disable recording mode for a given tape. The use of this function is
+       * mandatory if the auto-differentiable number is a taped type. However,
+       * for the purpose of developing generic code, it can also be safely
+       * called for tapeless auto-differentiable numbers.
+       *
+       * @note After this function call, the tape is considered ready for use and
+       * operations such as value(), gradient() or hessian() can be executed.
+       *
+       * @note This operation is only valid in recording mode.
+       */
+      void
+      disable_record_sensitivities(const bool write_tapes_to_file = false);
+
+      /**
+       * Select a pre-recorded tape to read from.
+       *
+       * @param[in] tape_index The index of the tape to be read from
+       *
+       * @note The chosen tape index must be greater than
+       * types::invalid_tape_index and less than types::max_tape_index.
```

Comment:
oh, are these numbers defined in namespace `types::`? we usually define these special numbers in namespace `numbers::` (e.g., `numbers::invalid_dof_index`)